SQLite Plugin

seancdavis opened the pull request:
Adds a SQLite source and destination. This was built from the Postgres plugin and doesn't differ much in terms of functionality.

SQLite is an oddball, however. The few important notes are mentioned [in the working doc](https://docs.google.com/document/d/165U4GtQTnvZC-NT_4VWVDC6DsKtdw5F58XoRIa6axn8/edit#) and will eventually be added to the website documentation.

Commit:
Copy @grouparoo/postgres to @grouparoo/sqlite
Without making any changes

Commit:
Configure sqlite app

Commit:
Can connect to sqlite database

Commit:
SQLite table source

Commit:
SQLite query source

Commit:
[WIP] Enabling SQLite destination

Commit:
[WIP] SQLite destination

Commit:
Profiles and groups are being exported!

Commit:
Support sqlite source file as a relative path

Commit:
Setup test utils for sqlite

Commit:
Specs for getChangedRowCount

Commit:
Move TODOs to the gdocs

Commit:
Pass table source specs

Commit:
Pass query source tests

Commit:
[WIP] Clean up int tests

Commit:
Add sqlite plugin to staging-community

Commit:
Remove lingering logs

Commit:
Remove pg packages

Commit:
Add a catchall for types since anything is technically a valid sqlite type

Commit:
Use more generic types for columns in tests

grouparoo-bot committed:
Updating things automatically

evantahler commented on plugins/@grouparoo/app-templates/src/source/table/templates.ts:
```diff
@@ -239,6 +239,12 @@ async function loadTablesAndColumns(
+        throw new Error(
+          'Could not find any listed columns in source. If using * make sure you used quotes ("*")?'
+        );
+      }
```
Note: This change helps avoid an error in bash/zsh in which an un-quoted `*` (vs `"*"`) would be expanded to match the files in this directory, and not be the string `"*"` like we wanted.

seancdavis replied:
This came in via #1502. Rebasing removed the change.

evantahler commented on plugins/@grouparoo/sqlite/__tests__/query-import/import-property.ts:
```diff
@@ -0,0 +1,101 @@
+    const value = await getPropertyValue(sql);
+    expect(value).toEqual(undefined);
+  });
+});
```
Because we know that SQLite has strange types, can you also add a test for a `date` here?  It may be helpful for the future to know the exact format we expect back, even if it is a string

evantahler commented on plugins/@grouparoo/sqlite/__tests__/table-import/get-changed-row-count.ts:
```diff
@@ -0,0 +1,69 @@
+    });
+  });
+
+  test("getChangedRowCount works", async () => {
+    const rowCount = await getChangedRowCount({
+      connection: client,
+      appOptions,
+      appId: app.id,
+      tableName: usersTableName,
+      highWaterMarkCondition: {
+        columnName: "stamp",
+        value: "2000/01/01 12:00:00",
+        filterOperation: FilterOperation.GreaterThan,
+      },
+    });
+
+    expect(rowCount).toBe(10);
+  });
```
Note: we are getting the string response from SQLite so the date formats should be OK

seancdavis replied:
I had updated the data formats to be consistent between the test data and the specs. Postgres was loading in dates like `2020/01/01`, but testing with a format like `2020-01-01`. Since we're working with strings, I chose to a consistent format. After making that change, everything worked as expected.

seancdavis replied:
Added some notes [to the doc](https://docs.google.com/document/d/165U4GtQTnvZC-NT_4VWVDC6DsKtdw5F58XoRIa6axn8/edit#).

evantahler commented on plugins/@grouparoo/sqlite/__tests__/table-import/import-properties.ts:
```diff
@@ -0,0 +1,881 @@
+            sourceMapping,
+            aggregationMethod: "min",
+          });
+          expect(values[profile.id]).toEqual(["2020/02/01"]);
+          expect(values[profile.id]).toEqual(["2020/02/01"]);
```
They remain strings... so it works!

evantahler commented on plugins/@grouparoo/sqlite/__tests__/table-import/import-property.ts:
```diff
@@ -0,0 +1,806 @@
+        );
+        expect(value).toBeGreaterThanOrEqual(0); // unpredictable ascii math
+      });
+      test("date", async () => {
+        const value = await getPropertyValue(
+          {
+            column,
+            sourceMapping,
+            aggregationMethod,
+          },
+          [{ op, key: "date", match: "2020/02/15" }]
+        );
+        expect(value).toEqual(2);
+      });
```
Anything wacky here with `>` and `<` on dates (that are actually strings).  Does SQLite do the right thing?

seancdavis replied:
It does as long as the format matches. We don't have that explicit test here, but I'd say the HWM tests cover that condition.

evantahler commented on plugins/@grouparoo/sqlite/__tests__/table-import/profiles.ts:
```diff
@@ -0,0 +1,180 @@
+      });
+      expect(page1.importsCount).toBe(4);
+      expect(page1.sourceOffset).toBe(0);
+      expect(Object.values(page1.highWaterMark)[0]).toMatch("2020/02/04"); // the times changes based on the TZ of the test server, but the date seems to be OK...
```
Good note!

evantahler commented on plugins/@grouparoo/sqlite/__tests__/table-import/schedule-options.ts:
```diff
@@ -0,0 +1,54 @@
+  test("gets list of columns that can handle highwatermark", async () => {
+    const columns = await getColumns();
+    const columnNames = columns.map((r) => r.key).sort();
+    // In SQLite, data is essentially either text or a number, so they could all
+    // technically be HWM columns.
```
It looks like you have chosen to filter out strings at least, which is a good step to making this list possible HWMs shorter.

seancdavis replied:
It's actually not filtering anything out. It's just that this list is coming from the purchases data, so there aren't as many columns.

evantahler commented on plugins/@grouparoo/sqlite/__tests__/utils/data.ts:
```diff
@@ -0,0 +1,142 @@
+  file: path.join(os.tmpdir(), `${config.sequelize.database}.sqlite`),
+};
+
+const appId = "app_31bb06e8-0a4e-49c3-ad42-545f2e8662e1";
```
FYI this can be anything, like `test_app` in the future

seancdavis replied:
I made the change anyways, just to simplify.

evantahler commented on plugins/@grouparoo/sqlite/package.json:
```diff
@@ -0,0 +1,49 @@
+  "author": "Grouparoo Inc <hello@grouparoo.com>",
+  "name": "@grouparoo/sqlite",
+  "description": "The Grouparoo SQLite Plugin",
+  "version": "0.2.8-alpha.1",
```
Be sure to bump this, and the `@grouparoo/core` version

evantahler commented on plugins/@grouparoo/sqlite/package.json:
```diff
@@ -0,0 +1,49 @@
+  "dependencies": {
+    "@grouparoo/app-templates": "0.2.8-alpha.1",
+    "@types/sqlite3": "^3.1.7",
+    "sqlite3": "^5.0.2"
```
Please remove the `^` and use exact versions

evantahler commented on plugins/@grouparoo/sqlite/src/lib/connect.ts:
```diff
@@ -0,0 +1,22 @@
+
+  // Don't allow connections unless the file already exists. This avoids
+  // in-memory storage and creating the file during instantiation.
+  if (!fs.existsSync(dbPath) || fs.lstatSync(dbPath).isDirectory()) return null;
```
Maybe this should `throw` an explicit message about the database file not being found?

evantahler commented on plugins/@grouparoo/sqlite/src/lib/sqlite.ts:
```diff
@@ -0,0 +1,48 @@
+      this.connection = new Database(this.database, (err: Error) => {
+        if (err) {
+          return reject(
+            new Error(`Error connecting to SQLite database: ${this.database}`)
```
The original error may be useful here too (perhaps append it to the string)?

evantahler commented on plugins/@grouparoo/sqlite/src/lib/sqlite.ts:
```diff
@@ -0,0 +1,48 @@
+    return new Promise((resolve, reject) => {
+      this.connection.all(query, (err: Error, res: any[]) => {
+        if (err) {
+          return reject(new Error(`Error running query: ${err}`));
```
I think you'll end up with `Error running query: Error: foo` I think.   Perhaps `new Error(`Error running query: ${err.message}`)`

seancdavis replied:
I ended up printing the error first and then the query second. Thought that could come in handy, as it's likely to be a syntax goof or something  I did.

evantahler commented on plugins/@grouparoo/sqlite/src/lib/test.ts:
```diff
@@ -0,0 +1,20 @@
+  // Return if we don't have a connection object.
+  if (!connection) return genericError;
+
+  const query = `SELECT count(*) FROM sqlite_master WHERE type = 'table' AND name != 'android_metadata' AND name != 'sqlite_sequence';`;
```
`android_metadata`?  Are there other strange os-specific tables we should be ignoring?

seancdavis replied:
I couldn't find a good list of this. But upon further inspection, anything that begins with `sqlite_` is reserved for internal use. Therefore, I've made the change to only filter out those tables beginning with `sqlite_` since any other table could be used for meta purposes, theoretically.

evantahler commented on plugins/@grouparoo/sqlite/src/lib/table-import/getColumns.ts:
```diff
@@ -0,0 +1,56 @@
+  FilterOperation.In,
+];
+
+const typeMap: { [columnType: string]: ColumnType } = {
+  bigint: "integer",
+  blob: "string",
+  boolean: "boolean",
+  date: "date",
+  datetime: "date",
+  email: "email",
+  float: "float",
+  int: "integer",
+  integer: "integer",
+  num: "float",
+  phoneNumber: "phoneNumber",
+  string: "string",
+  text: "string",
+  url: "url",
+  varchar: "string",
+};
```
We could be even fancier here, but for the demo, as long as the columns are named properly (eg: `email`), then the generated config files will be the right types.

seancdavis replied:
My thought exactly.

evantahler commented on plugins/@grouparoo/sqlite/src/lib/table-import/getSampleRows.ts:
```diff
@@ -0,0 +1,14 @@
+  connection,
+  tableName,
+}) => {
+  const query = `SELECT * FROM ${tableName} WHERE id IN (SELECT id FROM ${tableName} ORDER BY RANDOM() LIMIT 10)`;
```
Other databases got really slow when `ORDER BY RANDOM()` was used.  Is SQLite OK?

seancdavis replied:
I tested with 1k rows and the performance didn't bother me. Would you prefer I go the Postgres route and just pick the first 10?

evantahler replied:
I don't think it matters at the moment - but it really was a bottleneck for large MySQL/Postgres databases, so I wanted to keep an eye out for it

evantahler commented on plugins/@grouparoo/sqlite/tsconfig.json:
```diff
@@ -0,0 +1,14 @@
+    "outDir": "./dist",
+    "allowJs": true,
+    "module": "commonjs",
+    "target": "es2018",
```
we bumped to `es2019` since this branch was started

evantahler reviewed (changes requested):
Looking really good, and strong testing!  
I've added a few things to check and update.

Looks like you'll also need to do a rebase since the version release.

Commit:
bump es target

Commit:
Use specific package versions

Commit:
Add date spec to query property import

Commit:
Use more generic app_id when testing
Ignore only sqlite_* tables when querying for meta info
Some other small cleanup items

Commit:
Fix specs

Commit:
Fix in-memory error notice on connect

seancdavis commented:
We now have better connection errors!

<img width="1183" alt="Screen Shot 2021-03-24 at 6 58 24 AM" src="https://user-images.githubusercontent.com/5245089/112299787-84f6f380-8c6e-11eb-806b-cdbf9f41efa8.png">

<img width="1183" alt="Screen Shot 2021-03-24 at 6 58 47 AM" src="https://user-images.githubusercontent.com/5245089/112299794-86282080-8c6e-11eb-883f-8d23a2599dcb.png">

If using a relative path, the message will print the full path that it's looking for. It felt like it was more informative to do this, as opposed to being harmful and exposing the full working directory of the application. Especially since this tends to be an administrator-type of use case.
